src/lib.rs

```diff
@@ -2,67 +2,144 @@ use pyo3::{
-            children,
-            event_listeners,
-            state,
-            original_data: "".to_string(),
-            data: "".to_string(),
-            parent_element: None,
-            dom_pointer: None,
+        if f_name.ends_with(".fyre") {
+            filenames.push(absoltue_name);
         }
     }
 
-    pub fn is_text_node(&self) -> bool {
-        self.tag == "TEXT_NODE"
-    }
+    filenames
+}
+
+fn get_function_template(pyml_lines: &str, file_name: &str, root_name: &str) -> String {
+    let file_name = file_name.replace(".py", "");
+
+    if file_name.contains("__init__") {
+        return format!(
+            "
+from starfyre import create_component, render
+
+def fx_{0}():
+    # not nesting the code to preserve the frames
+    component = create_component(\"\"\"
+{1}
+        \"\"\")
+
+    return render(component)
+
+
+{0}=fx_{0}()
+    ",
+            root_name, pyml_lines
+        );
+    } else {
+        return format!(
+            "
+from starfyre import create_component
+
+def fx_{0}():
+    component = create_component(\"\"\"
+{1}
+        \"\"\")
+
+    return component
 
-    pub fn set_parent_element(&mut self, parent_element: Py<PyAny>) {
-        self.parent_element = Some(parent_element);
+
+{0}=fx_{0}()
+    ",
+            root_name, pyml_lines
+        );
     }
 }
 
-pub struct DomTree {
-    pub root: DomNode,
-}
+#[pyfunction]
+fn compile(file_name: &str) {
+    // get all the files in the directory of the file that end with .fyre
+    let fyre_files = get_all_fyre_files(file_name);
+
+    for fyre_file in fyre_files {
+        let python_file = fyre_file.replace(".fyre", ".py");
+
+        let fyre_file_pointer = File::open(fyre_file).unwrap();
+        let fyre_file_reader = std::io::BufReader::new(fyre_file_pointer);
+        let mut python_lines: Vec<String> = vec![];
+        let mut pyml_lines = vec![];
+
+        let mut python_flag = true;
+        let mut root_name = String::from("");
+
+        for line in fyre_file_reader.lines() {
+            let copy_line = line.unwrap().clone();
+            let is_pyml_line = copy_line.clone().trim().starts_with("<");
+            if !python_flag || is_pyml_line {
+                python_flag = false;
+                let copy_line = copy_line.clone();
+                if root_name == "" && is_pyml_line {
+                    let root_name_temp = copy_line.clone();
+                    // .split_once(' ').unwrap();
+                    let split: Vec<&str> = root_name_temp.split(' ').into_iter().collect();
+                    if !split.is_empty() && split[0] != "" {
+                        // check for trim
+                        root_name = split[0].replace("<", "").replace(">", "").to_string();
+                        continue;
+                    }
+                };
 
+                if copy_line.trim().starts_with(&format!("<{0}", root_name))
+                    || copy_line.trim().starts_with(&format!("</{0}", root_name))
```

Comment:
Dont think i understand the `</{0}"` case.

Reply:
Oh, it is the closing tag - e.g. `<Tag>` and `</Tag>`

Reply:
oh right, that would assume there is only whitespace before the close tag. Is it legal to have close tag on same line as content? `ends_with`?

Reply:
I wouldn't worry about it that much as it is a horrible way of parsing atm. The parser should ideally have a state machine of some sort to switch between pyml, CSS, and js. Then proper sanitization and then lexing.

This is more or less duct tape programming for now.